File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-213435-24827/analysis/acceptance-project/src/customerimporter/zip_code.py
from abc import abstractmethod

from customerimporter.typed_value import TypedValue


class ZipCode(TypedValue):

    # instance creation

    @classmethod
    def for_(cls, a_zip_code):
        return cls._type_representing(a_zip_code)(a_zip_code)

    # error messages

    @classmethod
    def invalid_type_error_description(cls):
        # the message the tests expect for a zip code that is neither old nor new
        return "Invalid identification type"

    # zip code

    @abstractmethod
    def code(self):
        pass

    # new zip code

    def has_new_zip_code(self):
        return False

    def new_zip_code_if_none(self, a_none_block):
        return a_none_block()

    # old zip code

    def has_old_zip_code(self):
        return False

    def old_zip_code_if_none(self, a_none_block):
        return a_none_block()

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-202259-53404/analysis/own-tests-project/src/main/java/customerimporter/ZipCode.java
package customerimporter;

import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

/**
 * The zip code of an {@link Address}. Each subclass recognizes the strings it can be created
 * from, validates them and knows its own value, so nobody has to ask which kind of zip code
 * it is dealing with.
 */
public abstract class ZipCode {

    private static final List<Kind> kinds = List.of(
        new Kind(OldZipCode::canBeCreatedFrom, OldZipCode::with),
        new Kind(NewZipCode::canBeCreatedFrom, NewZipCode::with));

    // instance creation

    public static ZipCode of(String aZipCode) {
        return kinds.stream()
            .filter(aKind -> aKind.handles(aZipCode))
            .findFirst()
            .orElseThrow(() -> new RuntimeException(invalidTypeErrorDescription()))
            .create(aZipCode);
    }

    // error messages

    // The tests expect this message for a zip code that is neither old nor new, so it is kept
    // as it was written even though it talks about identifications.
    public static String invalidTypeErrorDescription() {
        return "Invalid identification type";
    }

    // accessing

    public abstract Object value();

    // testing

    public abstract boolean isNew();

    public abstract boolean isOld();

    // evaluating

    public abstract Object newZipCodeIfNone(Supplier<Object> aNoneBlock);

    public abstract Object oldZipCodeIfNone(Supplier<Object> aNoneBlock);

    // kinds - private

    private record Kind(Predicate<String> condition, Function<String, ZipCode> creator) {

        boolean handles(String aZipCode) {
            return condition.test(aZipCode);
        }

        ZipCode create(String aZipCode) {
            return creator.apply(aZipCode);
        }
    }
}
